xcm/xcm-executor/src/lib.rs

```diff
@@ -204,24 +204,28 @@ impl<Config: config::Config> ExecuteXcm<Config::RuntimeCall> for XcmExecutor<Con
 			message,
 			weight_credit,
 		);
+		let mut properties = Properties { weight_credit, message_id: None };
 		if let Err(e) = Config::Barrier::should_execute(
 			&origin,
 			message.inner_mut(),
 			xcm_weight,
-			&mut weight_credit,
+			&mut properties,
 		) {
 			log::trace!(
 				target: "xcm::execute_xcm_in_credit",
-				"Barrier blocked execution! Error: {:?}. (origin: {:?}, message: {:?}, weight_credit: {:?})",
+				"Barrier blocked execution! Error: {:?}. (origin: {:?}, message: {:?}, properties: {:?})",
 				e,
 				origin,
 				message,
-				weight_credit,
+				properties,
 			);
+			// TODO: include `properties.message_id` in outcome?
 			return Outcome::Error(XcmError::Barrier)
 		}
 
-		let mut vm = Self::new(origin, message_hash);
+		*id = properties.message_id.unwrap_or(*id);
+
+		let mut vm = Self::new(origin, *id);
```

Comment:
here `id` can be possibly `message_hash` or `message_id vs topic` (depends on input and/or XcmConfig barrier setup):

1. [XcmContext](https://github.com/paritytech/polkadot/blob/master/xcm/src/v3/mod.rs#L339-L352) has attribute `message_hash: XcmHash`, which is now not very clear, for example if anybody wants to use this attribute with meaning of "real message hash", it could lead to possible bugs

2. iiuc, all barriers `AllowSetTopic / RequireSetTopic /  ExtractIdFromAppendedTopic` remove `SetTopic` instruction and sets `properties.message_id` (as topic). This extracted topic ends in `XcmContext.message_hash` [here](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-executor/src/lib.rs#L275-L280), but in this case `XcmContext.topic` is never used/set? So with this barriers there is no way to set `XcmContext.topic`?

Reply:
`AllowSetTopic` and `RequireSetTopic` are now removed.

`XcmContext` is used to inform various XCM APIs as to the current XCVM context. The `topic` field is set to whatever is in the current Topic register, which can be changed at will by the XCM through the `SetTopic` instruction.

The framework introduced here uses a trailing `SetTopic` instruction (which is essentially a no-op) to indicate the message's identity.

You're right that `XcmContext::message_hash` should be renamed to `XcmContext::message_id` to indicate the fact that it's not necessarily the message's hash. It was not done in order to preserve a bit of API stability, but I suppose it's fine to change.

Reply:
(`message_hash` is now `message_id`)

Reply:
ah, ok, thank you for explanation, got it